### ISSUE TITLE
racket-send-last-sexp after `#(1 2)` results in an error 

### ISSUE THREAD

**Issue by drdv** [greghendershott/racket-mode#734]
<details>
<h2>Package</h2>
<dl>
<dt>metadata</dt><dd><pre>(#s(package-desc racket-mode
		 (20240418 1446)
		 "Racket editing, REPL, and more"
		 ((emacs
		   (25 1)))
		 nil nil "/home/drdv/.emacs.d/elpa/racket-mode-20240418.1446"
		 ((:url . "https://www.racket-mode.com/")
		  (:maintainer "Greg Hendershott")
		  (:maintainers
		   ("Greg Hendershott"))
		  (:authors
		   ("Greg Hendershott" . "racket-mode-author@greghendershott.com"))
		  (:commit . "d5f4007fd2c2604673125035f4852dce1b639b33"))
		 nil))
</pre></dd>
<dt>package-archives</dt><dd><pre>(("gnu" . "https://elpa.gnu.org/packages/")
 ("nongnu" . "https://elpa.nongnu.org/nongnu/")
 ("melpa" . "https://melpa.org/packages/"))
</pre></dd>
<dt>racket--el-source-dir</dt><dd><pre>"/home/drdv/.emacs.d/elpa/racket-mode-20240418.1446/"</pre></dd>
<dt>racket--rkt-source-dir</dt><dd><pre>"/home/drdv/.emacs.d/elpa/racket-mode-20240418.1446/racket/"</pre></dd>
</dl>
<h2>System values</h2>
<dl>
<dt>emacs-version</dt><dd><pre>"29.3"</pre></dd>
<dt>major-mode</dt><dd><pre>help-mode</pre></dd>
<dt>system-type</dt><dd><pre>gnu/linux</pre></dd>
<dt>x-gtk-use-system-tooltips</dt><dd><pre>t</pre></dd>
<dt>display-graphic-p</dt><dd><pre>t</pre></dd>
</dl>
<h2>Buffer values</h2>
<dl>
<dt>after-change-functions</dt><dd><pre>nil</pre></dd>
<dt>before-change-functions</dt><dd><pre>nil</pre></dd>
<dt>completion-at-point-functions</dt><dd><pre>(tags-completion-at-point-function)
</pre></dd>
<dt>eldoc-documentation-function</dt><dd><pre>eldoc-documentation-default</pre></dd>
<dt>font-lock-defaults</dt><dd><pre>nil</pre></dd>
<dt>pre-command-hook</dt><dd><pre>(delete-selection-pre-hook which-key--hide-popup tooltip-hide)
</pre></dd>
<dt>post-command-hook</dt><dd><pre>(global-font-lock-mode-check-buffers global-eldoc-mode-check-buffers global-hl-todo-mode-check-buffers global-page-break-lines-mode-check-buffers ido-cr+-schedule-dynamic-collection-update mode-local-post-major-mode-change)
</pre></dd>
<dt>post-self-insert-hook</dt><dd><pre>(electric-pair-post-self-insert-function electric-pair-open-newline-between-pairs-psif electric-indent-post-self-insert-function blink-paren-post-self-insert-function)
</pre></dd>
<dt>xref-backend-functions</dt><dd><pre>(etags--xref-backend)
</pre></dd>
</dl>
<h2>Racket Mode values</h2>
<dl>
<dt>racket--cmd-open-p</dt><dd><pre>t</pre></dd>
<dt>racket-after-run-hook</dt><dd><pre>nil</pre></dd>
<dt>racket-before-run-hook</dt><dd><pre>(racket-ansi-color-context-reset)
</pre></dd>
<dt>racket-browse-url-function</dt><dd><pre>browse-url</pre></dd>
<dt>racket-command-timeout</dt><dd><pre>10</pre></dd>
<dt>racket-documentation-search-location</dt><dd><pre>"https://docs.racket-lang.org/search/index.html?q=%s"</pre></dd>
<dt>racket-error-context</dt><dd><pre>medium</pre></dd>
<dt>racket-expand-hiding</dt><dd><pre>standard</pre></dd>
<dt>racket-hash-lang-token-face-alist</dt><dd><pre>((constant . font-lock-constant-face)
 (error . error)
 (other . font-lock-doc-face)
 (keyword . font-lock-keyword-face)
 (hash-colon-keyword . racket-keyword-argument-face)
 (at . font-lock-doc-face))
</pre></dd>
<dt>racket-history-filter-regexp</dt><dd><pre>"\\`\\s *\\'"</pre></dd>
<dt>racket-imagemagick-props</dt><dd><pre>nil</pre></dd>
<dt>racket-images-inline</dt><dd><pre>t</pre></dd>
<dt>racket-images-keep-last</dt><dd><pre>100</pre></dd>
<dt>racket-images-system-viewer</dt><dd><pre>"display"</pre></dd>
<dt>racket-indent-curly-as-sequence</dt><dd><pre>t</pre></dd>
<dt>racket-indent-sequence-depth</dt><dd><pre>0</pre></dd>
<dt>racket-logger-config</dt><dd><pre>((cm-accomplice . warning)
 (GC . info)
 (module-prefetch . warning)
 (optimizer . info)
 (racket/contract . error)
 (racket-mode-debugger . info)
 (sequence-specialization . info)
 (* . fatal))
</pre></dd>
<dt>racket-memory-limit</dt><dd><pre>2048</pre></dd>
<dt>racket-module-forms</dt><dd><pre>"\\s(\\(?:module[*+]?\\|library\\)"</pre></dd>
<dt>racket-pretty-lambda</dt><dd><pre>nil</pre></dd>
<dt>racket-pretty-print</dt><dd><pre>t</pre></dd>
<dt>racket-program</dt><dd><pre>"racket"</pre></dd>
<dt>racket-repl-buffer-name-function</dt><dd><pre>nil</pre></dd>
<dt>racket-repl-command-file</dt><dd><pre>"/home/drdv/.emacs.d/racket-mode/repl.rkt"</pre></dd>
<dt>racket-repl-history-directory</dt><dd><pre>"~/.emacs.d/racket-mode/"</pre></dd>
<dt>racket-sexp-comment-fade</dt><dd><pre>0.5</pre></dd>
<dt>racket-shell-or-terminal-function</dt><dd><pre>racket-shell</pre></dd>
<dt>racket-show-functions</dt><dd><pre>(racket-show-pseudo-tooltip)
</pre></dd>
<dt>racket-smart-open-bracket-enable</dt><dd><pre>nil</pre></dd>
<dt>racket-submodules-to-run</dt><dd><pre>((test)
 (main))
</pre></dd>
<dt>racket-use-repl-submit-predicate</dt><dd><pre>nil</pre></dd>
<dt>racket-xp-add-binding-faces</dt><dd><pre>nil</pre></dd>
<dt>racket-xp-after-change-refresh-delay</dt><dd><pre>1</pre></dd>
<dt>racket-xp-highlight-unused-regexp</dt><dd><pre>"^[^_]"</pre></dd>
<dt>racket-xp-mode-lighter</dt><dd><pre>(:eval
 (racket--xp-mode-lighter))
</pre></dd>
<dt>racket-xp-mode-lighter</dt><dd><pre>(:eval
 (racket--xp-mode-lighter))
</pre></dd>
<dt>racket-xp-mode-lighter</dt><dd><pre>(:eval
 (racket--xp-mode-lighter))
</pre></dd>
<dt>racket-xp-mode-lighter</dt><dd><pre>(:eval
 (racket--xp-mode-lighter))
</pre></dd>
<dt>racket-xp-mode-lighter</dt><dd><pre>(:eval
 (racket--xp-mode-lighter))
</pre></dd>
<dt>racket-xp-mode-lighter</dt><dd><pre>(:eval
 (racket--xp-mode-lighter))
</pre></dd>
</dl>
<h2>Minor modes</h2>
<dl>
<dt>enabled</dt><dd><pre>((amx-mode)
 (auto-composition-mode)
 (auto-compression-mode)
 (auto-encryption-mode)
 (blink-cursor-mode)
 (buffer-read-only)
 (column-number-mode)
 (delete-selection-mode)
 (electric-indent-mode)
 (electric-pair-mode)
 (file-name-shadow-mode)
 (font-lock-mode)
 (global-eldoc-mode)
 (global-font-lock-mode)
 (global-hl-todo-mode)
 (global-page-break-lines-mode)
 (ido-ubiquitous-mode)
 (indent-tabs-mode)
 (isearch-fold-quotes-mode)
 (line-number-mode)
 (mouse-wheel-mode)
 (override-global-mode)
 (page-break-lines-mode)
 (semantic-minor-modes-format)
 (shell-dirtrack-mode)
 (show-paren-mode)
 (tooltip-mode)
 (transient-mark-mode)
 (which-key-mode)
 (windmove-mode)
 (xclip-mode))
</pre></dd>
<details><summary>Disabled minor modes</summary>
<dt>disabled</dt><dd><pre>((2C-mode)
 (abbrev-mode)
 (amx-debug-mode)
 (auto-fill-function)
 (auto-fill-mode)
 (auto-save-mode)
 (auto-save-visited-mode)
 (buffer-face-mode)
 (button-mode)
 (cl-old-struct-compat-mode)
 (comint-fontify-input-mode)
 (company-mode)
 (company-search-mode)
 (compilation-minor-mode)
 (compilation-shell-minor-mode)
 (completion-in-region-mode)
 (context-menu-mode)
 (cursor-face-highlight-mode)
 (dash-fontify-mode)
 (defining-kbd-macro)
 (diff-auto-refine-mode)
 (diff-minor-mode)
 (dired-hide-details-mode)
 (dired-isearch-filenames-mode)
 (dired-omit-mode)
 (dired-utils-format-information-line-mode)
 (display-line-numbers-mode)
 (eldoc-mode)
 (electric-layout-mode)
 (electric-quote-mode)
 (global-company-mode)
 (global-dash-fontify-mode)
 (global-display-line-numbers-mode)
 (global-linum-mode)
 (global-prettify-symbols-mode)
 (global-semantic-highlight-edits-mode)
 (global-semantic-highlight-func-mode)
 (global-semantic-show-parser-state-mode)
 (global-semantic-show-unmatched-syntax-mode)
 (global-semantic-stickyfunc-mode)
 (global-visual-line-mode)
 (golden-ratio-mode)
 (header-line-indent-mode)
 (hl-todo-mode)
 (horizontal-scroll-bar-mode)
 (hs-minor-mode)
 (ido-cr+-debug-mode)
 (ido-everywhere)
 (isearch-mode)
 (jit-lock-debug-mode)
 (linum-mode)
 (lock-file-mode)
 (lost-selection-mode)
 (markdown-live-preview-mode)
 (menu-bar-mode)
 (minibuffer-electric-default-mode)
 (next-error-follow-minor-mode)
 (outline-minor-mode)
 (overwrite-mode)
 (paragraph-indent-minor-mode)
 (paredit-mode)
 (prettify-symbols-mode)
 (racket-hash-lang-repl-mode)
 (racket-smart-open-bracket-mode)
 (racket-xp-mode)
 (read-extended-command-mode)
 (semantic-highlight-edits-mode)
 (semantic-highlight-func-mode)
 (semantic-mode)
 (semantic-show-parser-state-mode)
 (semantic-show-unmatched-syntax-mode)
 (semantic-stickyfunc-mode)
 (sh-electric-here-document-mode)
 (shell-highlight-undef-mode)
 (size-indication-mode)
 (tab-bar-history-mode)
 (tab-bar-mode)
 (temp-buffer-resize-mode)
 (text-scale-mode)
 (tool-bar-mode)
 (transient-resume-mode)
 (treesit-explore-mode)
 (treesit-inspect-mode)
 (undelete-frame-mode)
 (url-handler-mode)
 (use-hard-newlines)
 (vc-dir-git-mode)
 (view-mode)
 (visible-mode)
 (visual-line-mode)
 (window-divider-mode)
 (xref-etags-mode))
</pre></dd>
</details>
</dl>
</details>

Steps to reproduce: 

Placing `#(1 2 3)` at the beginning of a line and running `C-x C-e` results in:
```
application: not a procedure;
 expected a procedure that can be applied to arguments
  given: 1
```
if seems that `#` is somehow not taken into account. 

When I add a space in front of the hashtag, i.e., ` #(1 2 3)`, `C-x C-e` works just fine. In both cases C-c C-c` works as expected.

**Comment by greghendershott** [comment]
Interesting. Thanks for reporting!

I can reproduce this -- provided the expressions start at column 0, and are typed live into the buffer. (If I save the buffer, and load it, the expressions in column 0 do work, at least for me.)

Other examples include `#"byte string"`,  `#hash()`,  `#hasheq()`.

In all cases, the leading chars -- like `#`, `#hash`, and `#hasheq` -- are supposed to be given "prefix" char syntax. That causes `backward-sexp` and `backward-prefix-chars` to work, which `C-x C-e` uses.

I think I'd never noticed this because I'd always happened to use `C-x C-e` on indented expressions. (I most often would use this with expressions nestled in a `test` or `examples` submodule.)

Anyway it ought to work from column 0, also.

---

It looks like I can fix this by adding a function to `syntax-propertize-extend-region-functions`, to force it to start at the previous line. I'll test that more before committing...

**Comment by drdv** [comment]
Indeed reverting the buffer solves the problem until I edit the line (even if I introduce a space at the beginning and delete it the problem appears again). The diff of the outputs of `M-x describe-char` for the two cases is interesting:
```
6c6,8
<                syntax: ' 	which means: prefix
---
>                syntax: w 14	which means: word,
> 	  is the first character of a comment-start sequence,
> 	  is the second character of a comment-end sequence
20c22,23
<   syntax-table         (6)
---
> 
> [back]
```
Noe that `#` is not "the second character", it is the first one at least as I can see. And then there is the `prefix` vs. `word` that you mentioned as well. I just add this description for completeness, it seems that you are already on top of things :) ... thanks

**Comment by greghendershott** [comment]
Although most "char syntax" applies to single characters, Emacs has the notion of two character sequences for things like `/*` and `*/` in C/C++.

Racket has (rarely used) `#|` and `|#`, which explains IIUC what you're seeing there.

TL;DR `#` can start a reader shorthand sequence, it can be part of a comment delimiter, and it can be a plain old char in an identifier like `foo#bar`.  Which is why Racket Mode has to do more work to classify it.

**Comment by drdv** [comment]
I observe exactly the same problem with the new version. 
```
<dt>metadata</dt><dd><pre>(#s(package-desc racket-mode
		 (20241216 1334)
		 "Racket editing, REPL, and more."
		 ((emacs
		   (25 1)))
		 nil nil "/home/drdv/.emacs.d/elpa/racket-mode-20241216.1334"
		 ((:authors
		   ("Greg Hendershott" . "racket-mode-author@greghendershott.com"))
		  (:revdesc . "dd40f6220c4c")
		  (:commit . "dd40f6220c4cddffb7458e53e0424ca819abba9d")
		  (:url . "https://github.com/greghendershott/racket-mode"))
		 nil))
</pre></dd>
```

**Comment by greghendershott** [comment]
That is... difficult for me to explain. Unless maybe there's somehow some stale definition leftover from before you updated??

0. I restarted Emacs, and re-tested that the change fixes the problem for me.

1. If it's not too annoying for you: Could you please restart Emacs, then try again?

2. If the bug still occurs, could you please `M-x describe-variable`, enter `syntax-propertize-extend-region-functions`, and tell me the value. I'm expecting you'd see:

```
syntax-propertize-extend-region-functions is a variable defined in ‘syntax.el’.

Its value is
(syntax-propertize-wholelines racket-syntax-propertize-extend-region)
Local in buffer issue-734.rkt; global value is 
(syntax-propertize-wholelines)
```

**Comment by drdv** [comment]
Yes, when I am in a buffer with `racket-mode` on, I get the same value for `syntax-propertize-extend-region-functions` as you (modulo filename). I verified that the `racket-syntax-propertize-extend-region` function gets executed by simply issuing a message from it. So the setup is OK (and I did restart emacs). 

Maybe your modification has impact on some of the tests you made, but it has no impact of the problem I observe. That is, whether `racket-syntax-propertize-extend-region` returns the `cons` you form or returns `nil` (manually hard-coded by me in your code - which would be equivalent to removing `racket-syntax-propertize-extend-region`) the behaviour I observe is the same. I have replicated the problem on fedora 39 and on macos with a rather independent setup and by removing all other packages I use (just in case). To be more specific, consider the buffer (after just having reloaded it):
``` racket
#lang racket


#(1 2)
```
With point behind `)`, executing `C-x C-e` works fine. Now, I press `C-a` (to go to beginning of line) and `backspace` (to remove one of the blank lines above) followed by `C-e` (to go to the end of line). At that point, running `C-x C-e`, gives the error I described before. If now you repeat this process but instead of pressing `backspace` you press `return` to reopen the empty line, `C-x C-e` works fine again. This is one way I get the problem (there are others). If I remove the empty line above in a different way, things may work fine or not depending on what I do. 

It is clear that the correct properties are not assigned e.g., to `#` in some cases. It seems to me that to solve this problem, `racket-syntax-propertize` should be modified. For example, adding the matching rule 
``` elisp
((rx line-start ?#)
     (0 "'"))
```
solves the above problem (certainly it creates others :) ). Of course, this is meant to be a test, a sane rue should be added (but I know very little about `racket` and `racket-mode` to do a useful job at it).

**Comment by greghendershott** [comment]
That commit definitely fixes -- for me -- the scenario I think you originally reported:

> Placing `#(1 2 3)` at the beginning of a line and running `C-x C-e`

Did that at least improve, for you, too?

---

Having said that, yes, I can reproduce the other scenario you just reported: Move to beginning of line, press backspace.

I'm surprised that wouldn't also be handled by my commit. But it's not. I'll investigate that now...

**Comment by drdv** [comment]
For me the original problem I reported and the description above are the same thing (just my first post is less concrete). 

>> Placing `#(1 2 3)` at the beginning of a line and running `C-x C-e`

> Did that at least improve, for you, too?

It depends on what is meant by "placing". I played around with your commit and I couldn't find a case in which it makes a difference (in all cases I tested, when I have a problem without your commit, I have a problem with it as well - surely I have missed something).

**Comment by greghendershott** [comment]
OK. Thanks for all your time so far.

This is one of those issues where the effort to develop a perfect understanding of the bug seems disproportionate.

I am supposed to give some regexp rules to the `syntax-propertize-rules` macro, which writes some code for me. Then we have some inversion of control, where Emacs calls me at times of its choosing, and I'm supposed to call that macro-generated code.  So the context is a bit of a hairball.

I'm pretty sure the regular expressions are correct. They work when applied to the entire buffer, on initial load. They ought to work, AFAICT, when run from a position midway through the buffer. And they do -- except when something like `#(1 2)` starts in column zero.

So... I feel like the only reasonable tactic here is to play with extending the region -- i.e. making syntax propertize run from an earlier starting position.

---

I do now see a mistake, dammit, with my original attempt to extend the region. It was assuming that `(point)` and the `start` of the region are the same.

Instead:

```diff
modified   racket-common.el
@@ -81,13 +81,12 @@
 (defun racket-syntax-propertize-extend-region (beg end)
   "Value for hook `syntax-propertize-extend-region-functions'.
 
-Expects to be called after `syntax-propertize-wholelines',
-adjusting that result to start one line earlier. Ensures
-propertization of prefix syntax at BOL; see issue #734."
-  (let ((inhibit-field-text-motion t))
-    (let ((new-beg (line-beginning-position 0)))
-      (unless (eql new-beg beg)
-        (cons new-beg end)))))
+Ensure propertization of # prefix syntax at BOL. Although I don't
+understand exactly why, `syntax-propertize-wholelines' is
+insufficient. See issue #734."
+  (when (and (eq ?# (char-after beg))
+             (< (point-min) beg))
+    (cons (1- beg) end)))
```

With this:

- I can type fresh expressions like `#(1 2)` in column zero, and it's propertized correctly.
- And, I can move to BOL, and press backspace, and it remains propertized correctly.

Both scenarios you reported, work.

You mentioned you tried tweaking that function to return `nil`, before. Would you be willing to try the above change, and see what it does for you?

**Comment by drdv** [comment]
The patch works. I agree with your arguments.
Thanks

**Comment by greghendershott** [comment]
Thank you very much for trying that patch. I'm relieved it worked for you, too! I'll merge a commit with that.
